src/core/service.c

```diff
@@ -1250,23 +1271,102 @@ static int service_coldplug(Unit *u) {
         return 0;
 }
 
+static int connext_unix_harder(Service *s, const OpenFile *of, int ofd) {
+        union sockaddr_union addr = {
+                .un.sun_family = AF_UNIX,
+        };
+        socklen_t sa_len;
+        static const int socket_types[] = { SOCK_DGRAM, SOCK_STREAM, SOCK_SEQPACKET };
+        int r;
+
+        assert(s);
+        assert(of);
+        assert(ofd >= 0);
+
+        r = sockaddr_un_set_path(&addr.un, FORMAT_PROC_FD_PATH(ofd));
+        if (r < 0)
+                return log_unit_error_errno(UNIT(s), r, "Failed to set sockaddr for %s: %m", of->path);
+
+        sa_len = r;
+
+        for (size_t i = 0; i < ELEMENTSOF(socket_types); i++) {
+                _cleanup_close_ int fd = -EBADF;
+
+                fd = socket(AF_UNIX, socket_types[i] | SOCK_CLOEXEC, 0);
+                if (fd < 0)
+                        return log_unit_error_errno(
+                                        UNIT(s), errno, "Failed to create socket for %s: %m", of->path);
+
+                r = RET_NERRNO(connect(fd, &addr.sa, sa_len));
+                if (r >= 0)
+                        return TAKE_FD(fd);
+
+                if (r != -EPROTOTYPE)
+                        return log_unit_error_errno(
+                                        UNIT(s), r, "Failed to connect socket for %s: %m", of->path);
+        }
+
+        return log_unit_error_errno(UNIT(s), SYNTHETIC_ERRNO(EINVAL), "Failed to connect to socket sun_path %s sun_family %d: %m", addr.un.sun_path, addr.un.sun_family);
```

Comment:
- Please drop "%m" when `SYNTHETIC_ERRNO()` is used.
- The log message sounds something like debugging info...

`log_unit_error_errno(UNIT(s), SYNTHETIC_ERRNO(EINVAL), "Failed to connect socket for \"%s\".", of ->path);`